sys/kern/kern_proc.c

```diff
@@ -2501,6 +2501,58 @@ sysctl_kern_proc_auxv(SYSCTL_HANDLER_ARGS)
 	return (error != 0 ? error : error2);
 }
 
+/*
+ * Return the c18n stats block from the target process.  The pointer and
+ * length are stored in the process's auxv vector and have to be retrieved
+ * first.
+ */
+static int
+sysctl_kern_proc_c18n(SYSCTL_HANDLER_ARGS)
+{
+	int *name = (int *)arg1;
+	u_int namelen = arg2;
+	struct proc *p;
+	struct cheri_c18n_info info;
+	int error;
+	void *buffer;
+	ssize_t n;
+
+	if (namelen != 1)
+		return (EINVAL);
+
+	error = pget((pid_t)name[0], PGET_WANTREAD, &p);
+	if (error != 0 ||
+	    (p->p_flag & P_SYSTEM) != 0 ||
+	    SV_PROC_FLAG(p, SV_CHERI) == 0 ||
+	    p->p_c18n_info == NULL)
+		goto out;
+
+	/* Validate state, and if good, copy into buffer and then copy out. */
+	n = proc_readmem(curthread, p, (vm_offset_t)p->p_c18n_info, &info,
+	    sizeof(info));
+	if (n != sizeof(info) ||
+	    info.version != 1 ||
+	    info.stats_size == 0 ||
+	    info.stats_size > C18N_STATS_MAX_LEN) {
+		error = ENOEXEC;
+		goto out;
+	}
+
+	buffer = malloc(info.stats_size, M_TEMP, M_WAITOK);
+	n = proc_readmem(curthread, p, (__cheri_addr vm_offset_t)info.stats,
```

Comment:
You can use __CAP_CHECK to verify that the data is in range.  That macro should likely be altered to require that the capability be unsealed as well as tagged.  You should also check that it has load permission.

Reply:
`__CAP_CHECK` does require the capability to be tagged, which in this causes it to always fail because `info.stats` is always untagged.

I don't understand why we need to do this check though. Wouldn't the userspace just leak it's own memory if it sets a bad value?

Reply:
Why is info.stats untagged? That seems completely wrong.

Causing program secrets to be trivial available by sysctl seems like a bad idea.

Reply:
It seems that `proc_readmem` uses the `UIO_SYSSPACE` flag which does a `bcopynocap_c` underneath, stripping all tags.

Reply:
Sounds like we need an _c variant.

Reply:
I wonder whether we want to extend `uio_rw` with UIO_READ_CAP UIO_WRITE_CAP variants, so that we can honor it in both `uiomove_flags` and `uiomove_fromphys` (and all possible variations) without having to carry around an extra flag. I think when we set the UIO_READ/WRITE we already know whether we expect capabilities to be there and the current scheme should work fine in most cases as we don't preserve capabilities by default.